Update and expand the section on distributed hypertables (#555)
timescale/docs.timescale.com-content: using-timescaledb/distributed-hypertables.md

usage][getting-started-multi-node].

In most cases, using a distributed hypertable is similar to using a
regular hypertable, including inserting, querying, and altering
it. Thus, for basic usage, please review the [documentation for regular
hypertables][using-hypertables].

For inserts and queries, however, a distributed hypertable has unique
performance characteristics and there are also additional limitations
due to its distributed nature. For instance, query performance is
heavily dependent on the ability to *push down* work to data nodes,
which in turn ties into how data is partitioned across the nodes. If
push down is not possible, or the query does not involve many data
nodes, the query performance of a distributed hypertable will likely
be worse than that of a regular hypertable due to the additional
network overhead.

Note, also, that distributed hypertables can live alongside
non-distributed tables and other objects; in fact, no objects are
distributed by default. Interactions between distributed hypertables
and non-distributed objects might not have the expected behavior. For
instance, setting permissions on a distributed hypertable only works

> if the roles involved exist identically on all data nodes. Further,

Author: I wanted to put a reference, but I don't think we have a section for role management yet.

joins between a local table and a distributed hypertable requires
fetching the raw data from data nodes and performing the join locally.

>:WARNING: Distributed hypertables are not yet recommended for
production use. For more information, please review the current
[limitations][distributed-hypertable-limitations]. You can also
[contact us][contact] or join the #multinode channel in our [community
Slack][slack].

> ### Creating a Distributed Hypertable [](create)

Reviewer: Should this be ## or ###, @solugebefola ?  I think these should all just be one level below the top, so ### -> ##, and #### -> ###?  (But I remember with API there was something funny, so letting solar opine).

> Creating a distribute hypertable is very similar to creating a regular

Reviewer 2: ```suggestion
Creating a distributed hypertable is very similar to creating a regular
```

```

>:WARNING: If there are no data nodes configured for the database, you
>will get an error when executing `create_distributed_hypertable` and
>the distributed hypertable will not be created. Please review the
>section on [setting up multi-node][getting-started-multi-node]
>for information on how to add data nodes.

This creates a multi-dimensional distributed hypertable across all
data nodes, partitioned along `time` and `location`. With the given
partitioning configuration, each data node will be responsible for a
specific subset of the data based on the value of `location`, allowing
concurrent and parallel execution of a query across the data nodes.

You can now insert data into the distributed hypertable and the data
will automatically be routed to the correct data node.

#### Best Practices for Partitioning

Like a regular hypertable, a distributed hypertable needs to be
partitioned along a "time" dimension (e.g., a `timestamptz`
column). However, for most distributed workloads, we recommend
multi-dimensional partitioning with an additional "space" dimension
that consistently partitions the data over the data nodes, similar to
traditional

> [sharding](https://en.wikipedia.org/wiki/Shard_(database_architecture)). If

Reviewer 2: Make into a footer link?

your data set has a column called something similar to `customerID`,
`deviceID`, or `location` (as in the example above), and it figures
frequently in the `GROUP BY` clause of queries, then it is likely a
good candidate column for space partitioning.

With space partitioning, multi-row inserts can often be split across

> the data nodes for increased insert performance, and queries can be

Author: There's _always_ a chance that multi-row inserts can be split across data nodes; i.e., two rows could always go to different chunks on two different data nodes, even without space partitioning.

What I want to say here is that space partitioning makes it much more likely that you will hit multiple data nodes, in particular if you are doing in-order inserts. 

I will try to clarify.

parallelized by executing a subset of the query on each data
node. TimescaleDB contains many optimizations to take advantage of the
partitioning scheme to make best use of multi-node resources.

> While a distributed hypertable works with only time partitioning it

Reviewer 2: ```suggestion
When a distributed hypertable works with only time partitioning it
```

will likely lead to inserts on only one data node at a time. Assuming
in-order inserts, only one chunk gets written at a time and each chunk
is placed on a data node in round-robin fashion. Similarly, queries
cannot be optimally parallelized unless they cover as many chunk time
intervals as there are data nodes.

### Inserting into a Distributed Hypertable [](insert)

Just like a regular hypertable, it is important to batch data when
inserting into a distributed hypertable. Each insert statement is its
own transaction, and the overhead and cost of this transaction is good
to amortize over many rows of data. With a distributed hypertable, the
transaction has additional costs due to the coordination that needs to
happen across data nodes (e.g., two-phase commit protocol). There are

> two ways to insert data:

Reviewer 2: @mfreed the two parentheticals don't seem necessary.  The second parenthetical uses 'similar' twice in close proximity

- `INSERT`: the access node sets up a multi-row prepared statement on
  each data node and then splits the original insert statement across
  these sub-statements. Each prepared statement allows buffering of up
  to `timescaledb.max_insert_batch_size` number of rows (default 1000)
  before the statement is complete and thus written to the data
  node. Throughput can be tuned by setting a different batch size,
  although the max size is limited by the max number of parameters of
  a prepared statement (32767), and the number of columns in each row.
  For example, if a distributed hypertable has 10 columns, the max

>   insert batch size is capped at 3276 rows.

Reviewer 2: is this the max insert batch size _per data node_ or overall max?

Author: Per data node. I'll clarify.

- `COPY`: the access node switches each data node to COPY mode and
  then routes each row to the correct data node in a stream.
  
Out of these two options, `COPY` will typically deliver the best
performance, although it doesn't support features like conflict handling
(`ON CONFLICT DO...`).

### Querying a Distributed Hypertable [](select)

The query performance of a distribute hypertable depends heavily
on the ability of the planner to _push down_ computations to data
nodes. Without push down, the query will execute on the access node by
fetching the raw data from each data node and then performing any
necessary computations locally. Therefore, queries that do not lead to
push down and do not involve a lot of data nodes are unlikely to
perform better on a distributed hypertable compared to a regular

> hypertable. 

Reviewer 2: Switch to enhancement commentary vs. limitation?
> Therefore, queries that can take advantage of push down and involve many data nodes are more likely to see better performance on a distributed hypertable compared to a regular hypertable.

   data nodes, yielding a partial result that is sent back to the
   access node and finalized by combining all the partials from each
   data node. An example is computing the `max` temperature; each data
   node computes a local max and then the access node finalizes the
   result by computing the `max` over all the data nodes' `max`
   results:
   ```
   SELECT max(temperature) FROM conditions;
   ```
2. **FULL**: a full push down offloads all the computation to data
   nodes and the access node only appends the individual results. To
   enable full push down, the `GROUP BY` clause must include *all* the
   partition keys (dimension columns) or *only* the first space
   partition key. An example is computing the `max` temperature for
   each location (room1, room2, etc.):
   
   ```
   SELECT location, max(temperature) FROM conditions 
   GROUP BY location; 
   ``` 
   
   Given that we use `location` as a space partition, each data node
   can compute the max on its own distinct subset of the data.

>:TIP: For push down to work, the settings variable
>[`enable_partitionwise_aggregate`][partitionwise_agg_guc] must be set

> >to `TRUE`.

Author: Unfortunately, it is FALSE by default. I could still add a note about that.

> >to `TRUE`.

Reviewer: Sounds good.  A discussion later why FALSE is default =)

#### Limitations of Query Push Down

The query planner might not always be able to push down queries,
however, or it can only push down parts of it:

- If the query involves a region that saw a change in the partitioning
  configuration (e.g., the number of space partitions increased), then
  a full push down might not be possible when it otherwise would
  be. In our example distributed hypertable, the change in
  partitioning might lead to data for the same location existing in
  two chunks that are stored on different data nodes. The planner must
  therefore avoid *full* push down whenever the query covers the
  repartitioning boundary. A *partial* push down is still possible,
  however.
- If the query includes non-immutable functions and expressions, the
  planner cannot push down those parts of the query since they cannot
  be guaranteed to generate a consistent result across each data
  node. For instance, the [`random()`][random-func] function is
  [`VOLATILE`][volatility] since it depends on the current seed and

>   the state of the pseudo-random sequence. If the function would be

Reviewer 2: ```suggestion
  the state of the pseudo-random sequence. If the function were to be
```

Author: shouldn't it be "If the function _was_ to be.."?

partitioning.

With space partitioning, multi-row inserts can often be split across
the data nodes for increased insert performance, and queries can be
parallelized by executing a subset of the query on each data
node. TimescaleDB contains many optimizations to take advantage of the
partitioning scheme to make best use of multi-node resources.

When a distributed hypertable works with only time partitioning it
will likely lead to inserts on only one data node at a time. Assuming
in-order inserts, only one chunk gets written at a time and each chunk
is placed on a data node in round-robin fashion. Similarly, queries
cannot be optimally parallelized unless they cover as many chunk time
intervals as there are data nodes.

## Inserting into a Distributed Hypertable [](insert)

Just like a regular hypertable, it is important to batch data when
inserting into a distributed hypertable. Each insert statement is
often its own transaction, and the overhead and cost of this
transaction is good to amortize over many rows of data. With a
distributed hypertable, the transaction has additional costs due to
the coordination that needs to happen across data nodes (e.g.,

> two-phase commit protocol). In such cases, a single INSERT transaction

Reviewer 3: Not sure if this is intended to refer to the statement or the concept. IMHO, it's one of the two below:
```suggestion
two-phase commit protocol). In such cases, a single `INSERT` transaction
```

or 

```suggestion
two-phase commit protocol). In such cases, a single insert transaction
```

to the access node involving many rows of data is processed by the
access node, such that the access node (a) splits the input set into
several smaller batches of rows (with each batch having those rows
that belong to a specific data node based on the distributed
hypertable's partitioning), and then (b) writes each batch of rows to

> its corresponding data node.

Reviewer 3: This is a very long sentence. It will probably be easier to read if you split it up into several.

Author: This sentence was proposed by Mike so I will keep for now. I think it naturally breaks up based on (a) and (b) without requiring full stops.

> There are two ways to insert data to the access node (which similarly
> uses corresponding methods when interacting with its data nodes):

Reviewer 3: ```suggestion
There are two ways to insert data to the access node:
```

Author: Parenthesis was added based on feedback from Mike, so not reverting now.

- `INSERT`: the access node sets up a multi-row prepared statement on
  each data node and then splits the original insert statement across
  these sub-statements. The access node can buffer up to
  `timescaledb.max_insert_batch_size` number of rows (default 1000)
  per data node before a prepared statement's limit is reached and
  gets flushed to the data node. Thus, if there are 10000 rows in the

>   original INSERT statement and three data nodes with the default

Reviewer 3: ```suggestion
  original `INSERT` statement and three data nodes with the default
```

  insert batch size, the insert would roughly require three full
  batches per data node and a partial final batch.

  By tuning the insert batch size, throughput can be optimized. The
  maximum insert batch size is, however, limited by the maximum number
  of parameters allowed in a prepared statement (32767), and the
  number of columns in each row. For example, if a distributed
  hypertable has 10 columns, the max insert batch size is capped at
  3276 rows.

> - `COPY`: the access node switches each data node to COPY mode and

Reviewer 3: ```suggestion
- `COPY`: the access node switches each data node to `COPY` mode and
```

  then routes each row to the correct data node in a stream.

Out of these two options, `COPY` will typically deliver the best
performance, although it doesn't support features like conflict
handling (`ON CONFLICT` clause) that are used for [upserts][upserts].

## Querying a Distributed Hypertable [](select)

The query performance of a distribute hypertable depends heavily on

> the ability of the planner to _push down_ computations to data

Reviewer 3: Underscores are usually not used in printed text. Emphasis or boldface is used.
```suggestion
the ability of the planner to *push down* computations to data
```

Author: In markdown, this is one way to do emphasis.

nodes. Without push down, the query will execute on the access node by
fetching the raw data from each data node and then performing any
necessary computations locally. Therefore, queries that can take
advantage of push down and involve many data nodes are more likely to
see better performance on a distributed hypertable compared to a

> regular hypertable. In particular, an aggregate computation that down
> samples data is a good candidate for push down since it both

Reviewer 3: ```suggestion
regular hypertable. In particular, an aggregate computation that down-samples data is a good candidate for push down since it both
```

distributes the computational workload across data nodes and reduces
the amount of data that needs to be transferred in the result set.

> The planner can consider two types of push downs for aggregates:

Reviewer 3: ```suggestion
The planner can consider two types of push-downs for aggregates:
```

Author: I don't think this is correct, but would defer to, e.g., Mike or Solar.

Reviewer 2: 🤷 no idea.  I think the issue is that "push down" is a verb that is contorting to be a noun here?  Maybe:
>...can consider two methods to push down aggregates

?
I'm strangely ambivalent though

> 1. **FULL**: a full push down offloads all the computation to data

Reviewer 3: ```suggestion
1. **FULL**: a full push-down offloads all the computation to data
```

   nodes and the access node only appends the individual results. To
   enable full push down, the `GROUP BY` clause must include *all* the
   partition keys (dimension columns) or *only* the first space
   partition key. An example is computing the `max` temperature for

>    each location (room1, room2, etc.):

Reviewer 3: ```suggestion
   each location (`room1`, `room2`, etc.):
```

>    ```
>    SELECT location, max(temperature) FROM conditions
>    GROUP BY location;
>    ```

Reviewer 3: ```suggestion
   ```sql
   SELECT location, max(temperature) FROM conditions
   GROUP BY location;
   ```
```

  therefore avoid *full* push down whenever the query covers the
  repartitioning boundary. A *partial* push down is still possible,
  however.
- If the query includes non-immutable functions and expressions, the
  planner cannot push down those parts of the query since they cannot
  be guaranteed to generate a consistent result across each data
  node. For instance, the [`random()`][random-func] function is
  [`VOLATILE`][volatility] since it depends on the current seed and
  the state of the pseudo-random sequence. If the function was to be
  pushed down to each data node, it would not generate a valid
  pseudo-random sequence from the point of view of the access node
  that executes the query. Another example is the
  [`now()`][current_time] function to get the current transaction
  time. This is a [`STABLE`][volatility] function that depends on the
  current time zone setting on each node. TimescaleDB specifically
  optimizes the use of `now()`, however, by constifying the function
  on the access node, thus it is able to push down the resulting
  timestamp to the data nodes, but this is not the case for many

>   `STABLE` functions.

Reviewer 3: This seems to mixing several different concepts. Would be good to split it up into separate points, for example, one point about `now()` being optimized, one about handling of `VOLATILE` functions, and one about handling of `STABLE` functions.

Author: Which concepts are you thinking of? To me this is all about the ability to push down based on the volatility of a function.

Reviewer 3: The bullet point is talking about volatile function, stable functions, and now in different ways. It's quite hard to follow and understand what is relevant. I would suggest to break out the point about "now" being optimized and put it separately. I would also suggest to talk about volatile and stable functions separately.

Author: volitile and stable functions are both non-immutable, so it makes sense to talk about them together. The important thing here is that they are _not_ immutable, so in that case I think it makes more sense to just say they're both non-immutable.

Author: I removed mentioning stable and volatile and moved the optimization for now out of the bullet list.

  assumes that the function does not exist on the data nodes and
  therefore will not push it down.

## Data Node Management [](data-node-management)

When a distributed hypertable is created, it will use all the
available data nodes by default. It is possible, however, to use only
a subset of data nodes with a particular distributed hypertable. This
is useful to, e.g., tie a distributed hypertable to data nodes that
have a specific performance profile.

If additional data nodes are added to a distributed database, the data
nodes are not automatically associated with existing distributed
hypertables. Instead, you need to explicitly *attach* a data node
using [`attach_data_node`][attach_data_node]:

```sql
SELECT add_data_node('node3', host => 'dn3.example.com');
SELECT attach_data_node('node3', hypertable => 'conditions');
```

When attaching a data node, the partitioning configuration of the
distributed hypertable is also updated to account for the additional
data node (i.e., the number of space partitions is automatically increased to
match), unless the function parameter `repartition` is set to
`FALSE`. The updated partitioning configuration ensures that the data

> node will be able to take on newly created chunks.

Reviewer 3: It is not very clear that the new data node will only be used with new chunks, that is, it will not re-partition existing chunks.

Author: I am intentionally not mentioning it here since it is only a current limitation. We're mentioning elsewhere that this limitation exists.

  stream. `COPY` typically delivers better performance than insert
  statements, although it doesn't support features like conflict
  handling (`ON CONFLICT` clause) that are used for
  [upserts][upserts].

## Querying a Distributed Hypertable [](select)

The query performance of a distribute hypertable depends heavily on
the ability of the planner to _push down_ computations to data
nodes. Without push down, the query will execute on the access node by
fetching the raw data from each data node and then performing any
necessary computations locally. Therefore, queries that can take
advantage of push down and involve many data nodes are more likely to
see better performance on a distributed hypertable compared to a
regular hypertable. In particular, an aggregate computation that
down-samples data is a good candidate for push down since it both
distributes the computational workload across data nodes and reduces
the amount of data that needs to be transferred in the result set.

The planner can consider two types of push downs for aggregates:

1. **FULL**: a full push down offloads all the computation to data
   nodes and the access node only appends the individual results. To

>    enable full push down, the `GROUP BY` clause must include *all* the

Reviewer 2: ```suggestion
   enable a full push down, the `GROUP BY` clause must include *all* the
```

   partition key. An example is computing the `max` temperature for
   each location (`office_1`, `office_2`, etc.):

   ```sql
   SELECT location, max(temperature) FROM conditions
   GROUP BY location;
   ```

   Given that we use `location` as a space partition, each data node
   can compute the max on its own distinct subset of the data.

1. **PARTIAL**: if a full push down is not possible, the planner will
   do a partial push down instead. In this case, it offloads most of
   the computation to data nodes, yielding a partial result that is
   sent back to the access node and finalized by combining all the
   partials from each data node. An example is computing the `max`
   temperature; each data node computes a local max and then the
   access node finalizes the result by computing the `max` over all
   the data nodes' `max` results:

   ```sql
   SELECT max(temperature) FROM conditions;
   ```

>:TIP: For push down to work, the settings variable
>[`enable_partitionwise_aggregate`][partitionwise_agg_guc] must be set
>to `TRUE` on the access node (the default is `FALSE`).

Other types of push downs can include sorting operations, groupings,

> and joins. Push down for joins is currently not supported, however. To

Reviewer 2: Yah, here too, not sure if it's a verb or noun...it seems to be used interchangeably

```

Note that you cannot detach a data node that still holds data for the
hypertable. To be able to detach a data node, ensure all its data is
deleted first.

### Native Replication [](native-replication)

A distributed hypertable can be configured to write each chunk to
multiple data nodes in order to replicate data at the chunk
level. This *native replication* ensures that a distributed hypertable
is protected against data node failures and provides an alternative to
fully replicating each data node using streaming replication. With
native replication, the access node would continue to rely on
streaming replication, however.

>:WARNING: Native replication is currently under development and lacks
>functionality for a complete high-availability solution. Some
>functionality described in this section might not yet be implemented
>or only have a partial feature set. For instance, there is currently
>no mechanism implemented to re-replicate a chunk when a data node is

> >deleted due to a failure.

Author: Added this to a separate note below.

  chunk's data need to be replicated. When a chunk is compressed or
  decompressed, no additional replication is needed since it is a
  local operation on a data node that can be applied to each
  replica. With streaming replication, however, compressing or
  decompressing a chunk would require writing all the data changes to
  one or more standby replica nodes.
- **Query optimizations**: Native replication enables a number of
  interesting optimizations and features that go beyond just providing
  high availability (HA). For instance, when a chunk exists on
  multiple data nodes, the planner could balance queries across
  different chunk replicas for better resource utilization. Similarly,
  if the planner detects a non-responsive or slow data node, it could
  ensure that only responsive nodes are included in the query plan.

There are also disadvantages with native replication. For instance, if
a data node holds data that is not part of a distributed hypertable,
this data won't be replicated. In such situations, streaming
replication is recommended.

To configure native replication for a distributed hypertable, set the

> parameter [`replication_factor`][replication_factor] to a value

Reviewer 2: Not sure there is a link for this one at the footer?

```

Note that you cannot detach a data node that still holds data for the

> hypertable. To be able to detach a data node, ensure that either (1)
> all its data is deleted first, or (2) the data is replicated on other
> data nodes (see the next section on native replication).

Reviewer 3: ```suggestion
hypertable. To be able to detach a data node, ensure that either
all its data is deleted first or the data is replicated on other
data nodes (see the next section on native replication).
```

## Native Replication [](native-replication)

A distributed hypertable can be configured to write each chunk to
multiple data nodes in order to replicate data at the chunk
level. This *native replication* ensures that a distributed hypertable
is protected against data node failures and provides an alternative to
fully replicating each data node using streaming replication.

While data nodes require no additional setup to use native
replication, the access node continues to rely on streaming
replication, however. In this regard, the access node has similar
requirements for high availability as any regular PostgreSQL instance.

The number of data nodes a chunk is written to is determined by a
distributed hypertable's `replication_factor` (the default value is 1,
i.e., no native replication), and it can be set when calling
[`create_distributed_hypertable`][create_distributed_hypertable]:

> ```
> SELECT create_distributed_hypertable('conditions', 'time', 'location', replication_factor => 3);
> ```

Reviewer 3: Long line.
```suggestion
```sql
SELECT create_distributed_hypertable('conditions', 'time', 'location',
            replication_factor => 3);
```
```

Alternatively, the function
[`set_replication_factor`][set_replication_factor] can be used to
enable native replication on an existing distributed hypertable.

>:WARNING: Native replication is currently under development and lacks
>functionality for a complete high-availability solution. Some
>functionality described in this section is not yet implemented or
>might have only a partial feature set. For instance, there is
>currently no implementation for (re-)replicating existing chunks in
>the background after a data node failure or increase in replication
>factor. We therefore recommend keeping the replication factor set at
>the default value of 1, and instead use streaming replication on each
>data node.

Once enabled, native replication happens as part of normal inserts, by
writing each row to multiple data nodes, and therefore requires no

> additional mechanism for replicating newly inserted data. Existing
> chunks, however, which are not replicated according to the configured

Reviewer 3: ```suggestion
additional mechanism for replicating newly inserted data. Existing
chunks, which are not replicated according to the configured
```
